# HG changeset patch
# User Rik <rik@octave.org>
# Date 1398968163 25200
#      Thu May 01 11:16:03 2014 -0700
# Node ID d57a83f2d73e0c79c82080d9ab050ea004501189
# Parent  a18e223224c516ba6aca1cca601e3bc586f94ac3
Preserve image class in cdata (bug #41240).

* graphics.cc (convert_cdata): Add support for bool matrix input.

* imshow.m: Remove hack converting single, integer, bool formats to double.

diff --git a/libinterp/corefcn/graphics.cc b/libinterp/corefcn/graphics.cc
--- a/libinterp/corefcn/graphics.cc
+++ b/libinterp/corefcn/graphics.cc
@@ -861,16 +861,18 @@ convert_cdata (const base_properties& pr
   if (cdata.is_uint8_type ())
     CONVERT_CDATA_1 (uint8NDArray, uint8_, false);
   else if (cdata.is_uint16_type ())
     CONVERT_CDATA_1 (uint16NDArray, uint16_, false);
   else if (cdata.is_double_type ())
     CONVERT_CDATA_1 (NDArray, , true);
   else if (cdata.is_single_type ())
     CONVERT_CDATA_1 (FloatNDArray, float_, true);
+  else if (cdata.is_bool_type ())
+    CONVERT_CDATA_1 (boolNDArray, bool_, false);
   else
     error ("unsupported type for cdata (= %s)", cdata.type_name ().c_str ());
 
 #undef CONVERT_CDATA_1
 
   return octave_value (a);
 }
 
diff --git a/scripts/image/imshow.m b/scripts/image/imshow.m
--- a/scripts/image/imshow.m
+++ b/scripts/image/imshow.m
@@ -168,21 +168,16 @@ function h = imshow (im, varargin)
 
   nans = isnan (im(:));
   if (any (nans))
     warning ("Octave:imshow-NaN",
              "imshow: pixels with NaN or NA values are set to minimum pixel value");
     im(nans) = display_range(1);
   endif
 
-  ## This is for compatibility.
-  if (! (indexed || (true_color && isinteger (im))) || islogical (im))
-    im = double (im);
-  endif
-
   ## Clamp the image to the range boundaries
   if (! (true_color || indexed || islogical (im)))
     low = display_range(1);
     high = display_range(2);
     im(im < low) = low;
     im(im > high) = high;
   endif
 
